graveler/committed/tree/diff.go

```diff
@@ -0,0 +1,189 @@
+package tree
+
+import (
+	"bytes"
+
+	"github.com/treeverse/lakefs/graveler"
+)
+
+type IteratorValue struct {
+	record *graveler.ValueRecord
+	part   *Part
+}
+
+type diffIterator struct {
+	left          Iterator
+	right         Iterator
+	leftVal       *IteratorValue
+	rightVal      *IteratorValue
+	needLeftNext  bool
+	nextRightNext bool
+	currentVal    *graveler.Diff
+	err           error
+}
+
+const (
+	done = iota
+	sameParts
+	sameIdentities
+	sameKeys
+	needStartPartLeft
+	needStartPartRight
+	leftBeforeRight
+	rightBeforeLeft
+)
+
+func NewDiffIterator(left Iterator, right Iterator) graveler.DiffIterator {
+	return &diffIterator{
+		left:          left,
+		right:         right,
+		leftVal:       &IteratorValue{},
+		rightVal:      &IteratorValue{},
+		needLeftNext:  true,
+		nextRightNext: true,
+	}
+}
+
+func (d *diffIterator) next(it Iterator, val **IteratorValue) {
+	if it.Next() {
+		(*val).record, (*val).part = it.Value()
+		return
+	}
+	*val = nil
+	if it.Err() != nil {
+		d.err = it.Err()
+	}
+
+}
+
+func (d *diffIterator) nextPart(it Iterator, val **IteratorValue) {
+	if it.NextPart() {
+		(*val).record, (*val).part = it.Value()
+		return
+	}
+	*val = nil
+	if it.Err() != nil {
+		d.err = it.Err()
+	}
+}
+
+func key(it Iterator) []byte {
```

Comment:
func name should describe better what it does or at least include 'iterator' in the name so it will not collide.